libselinux/man/man3/selabel_digest.3
.TH "selabel_digest" "3" "16 Sept 2015" "" "SELinux API documentation"
.SH "NAME"
selabel_digest \- Return digest of specfiles and list of files used
.
.SH "SYNOPSIS"
.B #include <selinux/selinux.h>
.br
.B #include <selinux/label.h>
.sp
.BI "int selabel_digest(struct selabel_handle *" hnd ,
.in +\w'int selabel_digest('u
.BI "unsigned char **" digest ,
.BI "size_t *" digest_len ,
.br
.BI "char ***" specfiles,
.BI "size_t *" num_specfiles ");"
.in
.
.SH "DESCRIPTION"
.BR selabel_digest ()
performs an operation on the handle
.IR hnd ,
returning the results of the SHA256 digest pointed to by
.IR digest ,
whose length will be
.IR digest_len .
The list of specfiles used in the SHA256 digest calculation is returned in
.I specfiles
with the number of entries in
.IR num_specfiles .
.sp
To enable
.BR selabel_digest ()
to return this information the
.B SELABEL_OPT_DIGEST
option must be enable in
.BR selabel_open (3).
.sp
The result of
.BR selabel_digest ()
must not be used after
.BR selabel_close (3).
.
.SH "RETURN VALUE"
On success, zero is returned.  On error, \-1 is returned and
.I errno
is set appropriately.
.
.SH "ERRORS"
.TP
.B EINVAL
No digest available (returned if
.B SELABEL_OPT_DIGEST
option not enabled).
.TP
.B ENOMEM
An attempt to allocate memory failed.
.
.SH "SEE ALSO"
.BR selabel_open (3),
.BR selinux (8)


libsepol/ChangeLog
	* Fix memory leak in expand.c, from William Roberts.
	* Fix invalid read when policy file is corrupt, from William Roberts.
	* Fix possible use of uninitialized variables, from William Roberts.
	* Warn instead of fail if permission is not resolved, from James Carter.
	* Ignore object_r when adding userrole mappings to policydb, from Steve Lawrence.
	* Add missing return to sepol_node_query(), from Petr Lautrbach.
	* Correctly detect unknown classes in sepol_string_to_security_class, from Joshua Brindle.
	* Sort object files for deterministic linking order, from Laurent Bigonville.
	* Fix neverallowxperm checking on attributes, from Jeff Vander Stoep.
	* Remove libsepol.map when cleaning, from Nicolas Iooss.
	* Add high-level language line marking support to CIL, from James Carter.
	* Change logic of bounds checking to match change in kernel, from James Carter.
	* Fix multiple spelling errors, from Laurent Bigonville.
	* Only apply bounds checking to source types in rules, from Stephen Smalley.
	* Fix CIL and not add an attribute as a type in the attr_type_map, from James Carter
	* Build policy on systems not supporting DCCP protocol, from Richard Haines.
	* Fix extended permissions neverallow checking, from Jeff Vander Stoep.
	* Fix CIL neverallow and bounds checking, from James Carter
	* Android.mk: Add -D_GNU_SOURCE to common_cflags, from Nick Kralevich.
	* Add support for portcon dccp protocol, from Richard Haines
	* Fix bug in CIL when resetting classes, from Steve Lawrence

2.5 2016-02-23
	* Fix unused variable annotations, from Nicolas Iooss.
	* Fix uninitialized variable in CIL, from Nicolas Iooss.
	* Validate extended avrules and permissionxs in CIL, from Steve Lawrence.
	* Add support in CIL for neverallowx, from Steve Lawrence.
	* Fully expand neverallowxperm rules, from Richard Haines.
	* Add support for unordered classes to CIL, from Yuli Khodorkovskiy.
	* Add neverallow support for ioctl extended permissions, from Jeff Vander Stoep.
	* Improve CIL block and macro call recursion detection, from Steve Lawrence
	* Fix CIL uninitialized false positive in cil_binary, from Yuli Khodorkovskiy
	* Provide error in CIL if classperms are empty, from Yuli Khodorkovskiy
	* Add userattribute{set} functionality to CIL, from Yuli Khodorkovskiy
	* fix CIL blockinherit copying segfault and add macro restrictions, from Steve Lawrence
	* fix CIL NULL pointer dereference when copying classpermission/set, from Steve Lawrence
	* Add CIL support for ioctl whitelists, from Steve Lawrence
	* Fix memory leak when destroying avtab, from Steve Lawrence
	* Replace sscanf in module_to_cil, from Yuli Khodorkovskiy.
	* Improve CIL resolution error messages, from Steve Lawrence
	* Fix policydb_read for policy versions < 24, from Stephen Smalley.
	* Added CIL bounds checking and refactored CIL Neverallow checking, from James Carter
	* Refactored libsepol Neverallow and bounds (hierarchy) checking, from James Carter
	* Treat types like an attribute in the attr_type_map, from James Carter
	* Add new ebitmap function named ebitmap_match_any(), from James Carter
	* switch operations to extended perms, from Jeff Vander Stoep.
	* Write auditadm_r and secadm_r roles to base module when writing CIL, from Steve Lawrence
	* Fix module to CIL to only associate declared roleattributes with in-scope types, from Steve Lawrence
	* Don't allow categories/sensitivities inside blocks in CIL, from Yuli Khodorkovskiy.
	* Replace fmemopen() with internal function in libsepol, from James Carter.
	* Verify users prior to evaluating users in cil, from Yuli Khodorkovskiy.
	* Binary modules do not support ioctl rules, from Stephen Smalley.
	* Add support for ioctl command whitelisting, from Jeff Vander Stoep.
	* Don't use symbol versioning for static object files, from Yuli Khodorkovskiy.
	* Add sepol_module_policydb_to_cil(), sepol_module_package_to_cil(), and sepol_ppfile_to_module_package(), from James Carter.
	* Move secilc out of libsepol, from Yuli Khodorkovskiy.
	* fix building Xen policy with devicetreecon, and add devicetreecon
	  CIL documentation, from Richard Haines.
	* bool_copy_callback set state on creation, from Thomas Hurd.
	* Add device tree ocontext nodes to Xen policy, from Daniel De Graaf.
	* Widen Xen IOMEM context entries, from Daniel De Graaf.
	* Update CIL documentation, from Richard Haines
	* Fix error path in mls_semantic_level_expand(), from Chris PeBenito.
	* Fix MacOS X build, from Stephen Smalley.
	* Enabling building CIL in Android, from Stephen Smalley.
	* Update to latest CIL, includes new name resolution and fixes ordering
	  issues with blockinherit statements, and bug fixes

2.4 2015-02-02
	* Remove assumption that SHLIBDIR is ../../ relative to LIBDIR, from Steve
	  Lawrence
	* Fix bugs found by hardened gcc flags, from Nicolas Iooss.
	* Build CIL into libsepol. libsepol can be built without CIL by setting the
	  DISABLE_CIL flag to 'y', from Steve Lawrence
	* Add an API function to set target_platform, from Steve Lawrence
	* Report all neverallow violations, from Stephen Smalley
	* Improve check_assertions performance through hash tweaks from John Brooks.
	* Allow libsepol C++ static library on device from Daniel Cashman.

2.3 2014-05-06
	* Improve error message for name-based transition conflicts.
	* Revert libsepol: filename_trans: use some better sorting to compare and merge.
	* Report source file and line information for neverallow failures.
	* Fix valgrind errors in constraint_expr_eval_reason from Richard Haines.
	* Add sepol_validate_transition_reason_buffer function from Richard Haines.

2.2 2013-10-30
	* Allow constraint denial cause to be determined from Richard Haines.
	  - Add kernel policy version 29.
	  - Add modular policy version 17.
	  - Add sepol_compute_av_reason_buffer(), sepol_string_to_security_class(), sepol_string_to_av_perm().
	* Support overriding Makefile RANLIB from Sven Vermeulen.
	* Fix man pages from Laurent Bigonville.

2.1.9 2013-02-01

libselinux/utils/selabel_digest.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <errno.h>
#include <selinux/selinux.h>
#include <selinux/label.h>

static __attribute__ ((__noreturn__)) void usage(const char *progname)
{
	fprintf(stderr,
		"usage: %s -b backend [-v] [-B] [-i] [-f file]\n\n"
		"Where:\n\t"
		"-b  The backend - \"file\", \"media\", \"x\", \"db\" or "
			"\"prop\"\n\t"
		"-v  Run \"cat <specfile_list> | openssl dgst -sha256 -hex\"\n\t"
		"    on the list of specfiles to compare the SHA256 digests.\n\t"
		"-B  Use base specfiles only (valid for \"-b file\" only).\n\t"
		"-i  Do not request a digest.\n\t"
		"-f  Optional file containing the specs (defaults to\n\t"
		"    those used by loaded policy).\n\n",
		progname);
	exit(1);
}

static int run_check_digest(const char *cmd, const char *selabel_digest, size_t digest_len)
{
	FILE *fp;
	char files_digest[128];
	const char *files_ptr;
	int rc = 0;

	fp = popen(cmd, "r");
	if (!fp) {
		fprintf(stderr, "Failed to run command '%s':  %s\n", cmd, strerror(errno));
		return -1;
	}

	/* Only expect one line "(stdin)= x.." so read and find first space */
	while (fgets(files_digest, sizeof(files_digest) - 1, fp) != NULL)
		;

	files_ptr = strstr(files_digest, " ");

	rc = strncmp(selabel_digest, files_ptr + 1, digest_len * 2);
	if (rc) {
		printf("Failed validation:\n\tselabel_digest: %s\n\t"
				    "files_digest:   %s\n",
				    selabel_digest, files_ptr + 1);
	} else {
		printf("Passed validation - digest: %s\n", selabel_digest);
	}

	pclose(fp);
	return rc;
}

int main(int argc, char **argv)
{
	unsigned int backend = SELABEL_CTX_FILE;
	int rc, opt, validate = 0;
	char *baseonly = NULL, *file = NULL, *digest = (char *)1;
	char **specfiles = NULL;
	unsigned char *sha256_digest = NULL;
	size_t digest_len, i, num_specfiles;

	char cmd_buf[4096];
	char *cmd_ptr;
	char *sha256_buf = NULL;

	struct selabel_handle *hnd;
	struct selinux_opt selabel_option[] = {
		{ SELABEL_OPT_PATH, file },
		{ SELABEL_OPT_DIGEST, digest },
		{ SELABEL_OPT_BASEONLY, baseonly }
	};

	if (argc < 3)
		usage(argv[0]);

	while ((opt = getopt(argc, argv, "ib:Bvf:")) > 0) {
		switch (opt) {
		case 'b':
			if (!strcasecmp(optarg, "file")) {
				backend = SELABEL_CTX_FILE;
			} else if (!strcmp(optarg, "media")) {
				backend = SELABEL_CTX_MEDIA;
			} else if (!strcmp(optarg, "x")) {
				backend = SELABEL_CTX_X;
			} else if (!strcmp(optarg, "db")) {
				backend = SELABEL_CTX_DB;
			} else if (!strcmp(optarg, "prop")) {
				backend = SELABEL_CTX_ANDROID_PROP;
			} else if (!strcmp(optarg, "service")) {
				backend = SELABEL_CTX_ANDROID_SERVICE;
			} else {
				fprintf(stderr, "Unknown backend: %s\n",
								    optarg);
				usage(argv[0]);
			}
			break;
		case 'B':
			baseonly = (char *)1;
			break;
		case 'v':
			validate = 1;
			break;
		case 'i':
			digest = NULL;
			break;
		case 'f':
			file = optarg;
			break;
		default:
			usage(argv[0]);
		}
	}

	memset(cmd_buf, 0, sizeof(cmd_buf));

	selabel_option[0].value = file;
	selabel_option[1].value = digest;
	selabel_option[2].value = baseonly;

	hnd = selabel_open(backend, selabel_option, backend == SELABEL_CTX_FILE ? 3 : 2);
	if (!hnd) {
		switch (errno) {
		case EOVERFLOW:
			fprintf(stderr, "ERROR Number of specfiles or specfile"
					" buffer caused an overflow.\n");
			break;
		default:
			fprintf(stderr, "ERROR: selabel_open: %s\n",
						    strerror(errno));
		}
		return -1;
	}

	rc = selabel_digest(hnd, &sha256_digest, &digest_len, &specfiles,
							    &num_specfiles);

	if (rc) {
		switch (errno) {
		case EINVAL:
			fprintf(stderr, "No digest available.\n");
			break;
		default:
			fprintf(stderr, "selabel_digest ERROR: %s\n",
						    strerror(errno));
		}
		goto err;
	}

	sha256_buf = malloc(digest_len * 2 + 1);
	if (!sha256_buf) {
		fprintf(stderr, "Could not malloc buffer ERROR: %s\n",
						    strerror(errno));
		rc = -1;
		goto err;
	}

	printf("SHA256 digest: ");
	for (i = 0; i < digest_len; i++)
		sprintf(&(sha256_buf[i * 2]), "%02x", sha256_digest[i]);

	printf("%s\n", sha256_buf);
	printf("calculated using the following specfile(s):\n");

	if (specfiles) {
		size_t cmd_rem = sizeof(cmd_buf);
		int ret;

		if (validate) {
			cmd_ptr = &cmd_buf[0];
			ret = snprintf(cmd_ptr, cmd_rem, "/usr/bin/cat ");
			if (ret < 0 || (size_t)ret >= cmd_rem) {
				fprintf(stderr, "Could not format validate command\n");
				rc = -1;
				goto err;
			}
			cmd_ptr += ret;
			cmd_rem -= ret;
		}

		for (i = 0; i < num_specfiles; i++) {
			if (validate) {
				ret = snprintf(cmd_ptr, cmd_rem, "%s ", specfiles[i]);
				if (ret < 0 || (size_t)ret >= cmd_rem) {
					fprintf(stderr, "Could not format validate command\n");
					rc = -1;
					goto err;
				}
				cmd_ptr += ret;
				cmd_rem -= ret;
			}

			printf("%s\n", specfiles[i]);
		}

		if (validate) {
			ret = snprintf(cmd_ptr, cmd_rem, "| /usr/bin/openssl dgst -sha256 -hex");
			if (ret < 0 || (size_t)ret >= cmd_rem) {
				fprintf(stderr, "Could not format validate command\n");
				rc = -1;
				goto err;
			}

			rc = run_check_digest(cmd_buf, sha256_buf, digest_len);
		}
	}

err:
	free(sha256_buf);
	selabel_close(hnd);
	return rc;
}
